src/wallet/coin_selection.rs

```diff
@@ -201,45 +202,111 @@ impl<D: Database> CoinSelectionAlgorithm<D> for LargestFirstCoinSelection {
                 .chain(optional_utxos.into_iter().rev().map(|utxo| (false, utxo)))
         };
 
-        // Keep including inputs until we've got enough.
-        // Store the total input value in selected_amount and the total fee being paid in fee_amount
-        let mut selected_amount = 0;
-        let selected = utxos
-            .scan(
-                (&mut selected_amount, &mut fee_amount),
-                |(selected_amount, fee_amount), (must_use, weighted_utxo)| {
-                    if must_use || **selected_amount < amount_needed + **fee_amount {
-                        **fee_amount +=
-                            fee_rate.fee_wu(TXIN_BASE_WEIGHT + weighted_utxo.satisfaction_weight);
-                        **selected_amount += weighted_utxo.utxo.txout().value;
-
-                        log::debug!(
-                            "Selected {}, updated fee_amount = `{}`",
-                            weighted_utxo.utxo.outpoint(),
-                            fee_amount
-                        );
-
-                        Some(weighted_utxo.utxo)
-                    } else {
-                        None
-                    }
-                },
-            )
-            .collect::<Vec<_>>();
+        select_sorted_utxso(utxos, fee_rate, amount_needed, fee_amount)
+    }
+}
 
-        let amount_needed_with_fees = amount_needed + fee_amount;
-        if selected_amount < amount_needed_with_fees {
-            return Err(Error::InsufficientFunds {
-                needed: amount_needed_with_fees,
-                available: selected_amount,
-            });
+/// Another simple and dumb coin selection
+///
+/// This coin selection algorithm sorts the available UTXOs by blockheight and then picks them starting
+/// from the oldest ones until the required amount is reached.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct OldestFirstCoinSelection;
+
+impl<D: Database> CoinSelectionAlgorithm<D> for OldestFirstCoinSelection {
+    fn coin_select(
+        &self,
+        database: &D,
+        required_utxos: Vec<WeightedUtxo>,
+        mut optional_utxos: Vec<WeightedUtxo>,
+        fee_rate: FeeRate,
+        amount_needed: u64,
+        fee_amount: u64,
+    ) -> Result<CoinSelectionResult, Error> {
+        use bitcoin::hash_types::Txid;
+
+        let mut txids: Vec<Txid> = optional_utxos
+            .iter()
+            .map(|wu| wu.utxo.outpoint().txid)
+            .collect::<Vec<Txid>>();
+
+        txids.dedup();
+
+        // query db and create a blockheight lookup table
+        let mut blockheights: HashMap<Txid, u32> = HashMap::new();
```

Comment:
I think you could create this map without having to allocate an extra `Vec` for the txids: you could try mapping directly to a `(Txid, u32)` tuple by making the db lookup inside the closure. Then you can `collect()` into an `HashMap<Txid, u32>`.

Since errors could happen during the lookup, wrap everything in a `Result`, something like:

```rust
let blockheights = optional_utxos
    .iter()
    .map(|wu| wu.utxo.outpoint().txid)
    .filter_map(|txid| Ok(database.get_tx(&txid, false)?.and_then(|details| details.confirmation_time).map(|conf| (txid, conf.height))))
    .collect::<Result<HashMap<_, _>, Error>>()?;
```

Reply:
My original ideal of having a separate txids(and perform dedup) is to avoid duplicated db lookup if there are multiple utxos from same tx. But you are right that there is no need to having allocate an extra Vec for it . We can surely do it in 1 iterator

Your example result in some compilation error. but I am able to get it working in the below code

```
        let blockheights = optional_utxos
            .iter()
            .map(|wu| wu.utxo.outpoint().txid)
            .map(|txid| {
                database
                    .get_tx(&txid, false)
                    .map(|detail| {
                        detail
                        .and_then(|detail| detail.confirmation_time)
                        .map(|conf| (txid, conf.height))
                    })
            })
            .filter_map(|x| x.transpose())
            .collect::<Result<HashMap<_, _>, Error>>()?;
```

However, my latest implementation is using fold as I believed that's the only way that we can skip db lookup in pure iterator approach? Please let me know if u find any issue in the latest approach? Thank you very much for your thorough review

Reply:
Yes, I really like this approach!